bundles/org.openhab.binding.upb/src/main/java/org/openhab/binding/upb/handler/SerialPIMHandler.java

```diff
@@ -0,0 +1,176 @@
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.upb.handler;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.smarthome.core.thing.Bridge;
+import org.eclipse.smarthome.core.thing.ThingStatus;
+import org.eclipse.smarthome.core.thing.ThingStatusDetail;
+import org.eclipse.smarthome.io.transport.serial.PortInUseException;
+import org.eclipse.smarthome.io.transport.serial.SerialPort;
+import org.eclipse.smarthome.io.transport.serial.SerialPortIdentifier;
+import org.eclipse.smarthome.io.transport.serial.SerialPortManager;
+import org.eclipse.smarthome.io.transport.serial.UnsupportedCommOperationException;
+import org.openhab.binding.upb.Constants;
+import org.openhab.binding.upb.internal.message.MessageBuilder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Bridge handler responsible for serial PIM communications.
+ *
+ * @author Marcus Better - Initial contribution
+ *
+ */
+@NonNullByDefault
+public class SerialPIMHandler extends PIMHandler {
+    private static final int SERIAL_RECEIVE_TIMEOUT = 100;
+    private static final int BAUD_RATE = 4800;
+    private static final int SERIAL_PORT_OPEN_INIT_DELAY_MS = 500;
+    private static final int SERIAL_PORT_OPEN_RETRY_DELAY_MS = 30_000;
+
+    private final Logger logger = LoggerFactory.getLogger(SerialPIMHandler.class);
+
+    private SerialPortManager serialPortManager;
+    private volatile @Nullable SerialIoThread receiveThread;
+    private volatile @Nullable ScheduledFuture<?> futSerialPortInit;
+
+    public SerialPIMHandler(final Bridge thing, final SerialPortManager serialPortManager) {
+        super(thing);
+        this.serialPortManager = serialPortManager;
+    }
+
+    @Override
+    public void initialize() {
+        logger.debug("Initializing Serial UPB PIM {}.", getThing().getUID());
+        super.initialize();
+
+        final String portId = (String) getConfig().get(Constants.CONFIGURATION_PORT);
+        if (portId == null || portId.isEmpty()) {
+            logger.warn("serial port is not set");
+            return;
+        }
+
+        futSerialPortInit = scheduler.schedule(() -> openSerialPort(portId), SERIAL_PORT_OPEN_INIT_DELAY_MS,
+                TimeUnit.MILLISECONDS);
+    }
+
+    @Override
+    public void dispose() {
+        if (futSerialPortInit != null) {
+            futSerialPortInit.cancel(true);
+        }
+        futSerialPortInit = null;
```

Comment:
For example to solve this type of null warning you need to cache the field value into a local variable:
```suggestion
        ScheduledFuture<?> futSerialPortInit = this.futSerialPortInit;
        if (futSerialPortInit != null) {
            futSerialPortInit.cancel(true);
        }
        this.futSerialPortInit = null;
```

Yes it may seem unnecessary to do it like this but the warnings help prevent NPE in a concurrent environment so you should treat what it complains about very seriously with the exception of `Map` fields which the static analysis seems to have trouble with. If the warning relate to your usage of a `Map` field just ignore it for now, I'll help you address those later.

Reply:
Ok, done. It helped clean up the code in some cases.